_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+  hardware.
+- Don't assert on the contents of data read from persistent storage, unless it's
+  guaranteed to be valid. The data read from flash or a filesystem could be
+  corrupted.
+- If it is very likely the system will recover in a few moments, it might be
+  best to not assert.
+
+## Custom Assert Implementation Examples
+
+It's now time to start asserting ourselves. The examples provided below progress
+from a naive solution to the best one I've seen and used to date, with each
+becoming more expressive while also using less (precious) code space.
+
+> NOTE: All of these implementations will behave similarly if a debugger is
+> attached. The goal of the exercise is to improve upon asserts that are hit
+> while a debugger is **not** attached and all we have is a logging system.
+
+The code starts out in `prv_main`, and will call one of two functions,
+`prv_one_true`, or `prv_one_false`, depending on a randomly generated boolean
+variable. Both of functions will then call either `two_true` or `two_false`,
+which gives us **four** unique paths our code can take on boot.
+
+We are taking a random code path because we want to come up with an assert
+implementation that can tell us which path through the four functions our
+session took.
+
+### Setup
+
+For this setup we will use:
+
+- a nRF52840-DK[^5] (ARM Cortex-M4F) as our development board
+- SEGGER JLinkGDBServer[^6] as our GDB Server.
+- GCC 8.3.1 / GNU Arm Embedded Toolchain as our compiler[^7]
+- GNU make as our build system
+
+All the code can be found on the
+[Interrupt Github page](https://github.com/memfault/interrupt/tree/master/example/debugging-asserts)
+with more details in the `README` in the directory linked.
+
+#### Compiling the code and launching it with GDB
+
+```
+# Compile the first assert implementation
+$ IMPL=ASSERT1 make
+Compiling main.c
+Compiling startup.c
+Compiling freertos_kernel/tasks.c
+Compiling freertos_kernel/queue.c
+Compiling freertos_kernel/list.c
+Compiling freertos_kernel/timers.c
+Compiling freertos_kernel/portable/GCC/ARM_CM4F/port.c
+Compiling freertos_kernel/portable/MemMang/heap_1.c
+Compiling impls/assert1.c
+Linking library
+Generated build/nrf52.elf
+
+# In one terminal, start a GDB Server
+$ JLinkGDBServer  -if swd -device nRF52840_xxAA
+SEGGER J-Link GDB Server V6.52a Command Line Version
+
+# Flash the code on the NRF52 and start gdb
+$ arm-none-eabi-gdb-py --eval-command="target remote localhost:2331" --ex="mon reset" --ex="load"
+--ex="mon reset" --se=build/nrf52.elf
+GNU gdb (GNU Tools for Arm Embedded Processors 8-2019-q3-update) 8.3.0.20190703-git
+Copyright (C) 2019 Free Software Foundation, Inc.
+[...]
+Resetting target
+Loading section .interrupts, size 0x40 lma 0x0
+Loading section .text, size 0x194d lma 0x40
+Loading section .data, size 0x4 lma 0x1990
+Start address 0x40, load size 6545
+Transfer rate: 2130 KB/sec, 2181 bytes/write.
+Resetting target
+(gdb)
+```
+
+#### Assert Metadata Structure
```

Comment:
You should put this after the "logging assert" and "standard library assert" approach.

Reply:
We actually use this structure in all assert handlers after my changes.